### ISSUE TITLE
What term should we use for a federated gateway / router in this specification?

### ISSUE THREAD

**Issue by martijnwalraven** [graphql/composite-schemas-spec#17]
The draft specification in this repository currently refers to a *gateway configuration* for what we call a *supergraph schema* at Apollo. That opens up the question of how we should refer to the component that is responsible for executing federated queries.

At Apollo, we've started referring to this component as a *graph router* because we feel that better reflects the fact that it intelligently routes requests to appropriate subgraph servers as needed. But perhaps what is needed here is a more neutral term that describes its role within the architecture.

**Comment by michaelstaib** [comment]
I have thought about this a bit and it would be great to align with the graphql spec ... which does not mention any of this... rather it talks about an execution/executor.

> To execute a request, the executor must have a parsed [Document](https://spec.graphql.org/June2018/#Document) and a selected....

So we could call it a distributed executor/composite executor/federated executor.... something in that line.

**Comment by martijnwalraven** [comment]
> So we could call it a distributed executor/composite executor/federated executor.... something in that line.

Yeah, something along those lines makes sense to me.

**Comment by michaelstaib** [comment]
Dispatcher sounds more like a messaging system.

My thinking at the moment is that we are describing an alternate executor for GraphQL that gets is execution instructions from the composite schema. But I get where you are going at. In general an implementation of this is more doing the execution planing and then delegates the execution... basically plan -> dispatch -> merge.

**Comment by michaelstaib** [comment]
Orchestrator?

**Comment by benjie** [comment]
"Coordinator"?

**Comment by smyrick** [comment]
Quite relevant, I just made this comment on the other issue discussing what we call the `schemas`, where is this issues is asking what should we call the `servers`

tl;dr What about `supergraph server` and `supergraph schema` with `subgraph server` and `subgraph` schema?

https://github.com/graphql/composite-schemas-spec/issues/11#issuecomment-1945031531

**Comment by smyrick** [comment]
We did a brainstorm on the meeting on Feb 15 and came up with some more terms

- Distributed executor
- Supergraph resolver
- Composer (probably not good)
- Director (from composer making music)
- Conductor
- GraphQL API Gateway
- Coordinator

**Comment by JohnStarich** [comment]
My takes:
* Orchestrator - Interesting, a tad non-specific though. Not a bad option in my opinion.
* Coordinator - I like this. It's a little closer to a role description.
* Distributed executor – I like this one, it fully describes the role. The downside is its a bit verbose. "Execution distributor" might be slightly closer.
* Supergraph resolver / Resolver – This one's interesting. It borrows a term from the GraphQL spec to describe turning requests into responses, only now it's at a service rather than field level.
* Composer – Consensus was this name is already shared with other terms in the spec. It's also doing more than composition, since it is breaking up requests and sending them off to be executed.
* Director – This one is a bit clearer about its behavior. Director is a general term used to describe management of others, but usually to fulfill a shared goal.
* Conductor – Similar to Director, though possibly more metaphor-like, which may be undesirable.
* GraphQL API Gateway / Gateway – This has some precedence, and it's also a borrowed term from elsewhere like an "ingress gateway". I like this because it describes an entrypoint to a larger system. The only thing is it doesn't describe its behavior in the name.

Of these, my favorites in order are Director, Gateway, Conductor.

**Comment by benjie** [comment]
I still like "dispatcher" - did you discuss that today? I don't see it among those listed.

"Execution distributor" is also an interesting term; it splits up the execution and distributes it to the source schemas... I like it. Other than verbosity, I like it quite a lot - nice one :+1: 

I think we should avoid "resolver" in order to avoid confusion with GraphQL field resolvers.

Director and conductor are okay, I quite like them, but they generally are "self driven" (there's not really an "input", though arguably the piece of music itself is the input for a conductor) whereas a dispatcher performs a similar role but is driven by external input (request comes in, dispatcher splits it up and delegates it to the relevant parties and reports back).

**Comment by JohnStarich** [comment]
Ah, we must’ve missed Dispatcher. That’s a good one, I like it.

**Comment by PascalSenn** [comment]
There are similar concepts in other areas of CS, i'll just add the following to the discussion to think about: 

**Coordinator** - Distributed Transactions:
In distributed transactions there is a component called the coordinator that coordinates the transactions between different systems and manages the state of the transaction. 
I am not sure if the coordinator from distributed transactions is a good analogy in this case, as the two and three phase commits are really a fundamentally different concept with different communication patterns.

**Orechstrator** - Saga Pattern
In the saga pattern, specifically in the Orchestration saga, there is a component called the orchestrator.
Looking at sagas i see a lot of similarities. On the start of the execution a saga has a predefined flow/statemachine. The orchestrator executes this saga based on this plan. Usually this plan is hardcoded in the saga pattern, but in our case i guess it's the compose that create the plan that then is executed by the orechstrator. 

**(Query) Executor** - DBMS 
In Database system a incoming query is compiled into an **execution plan** that is the handed over to the *Query Executor* that executes the request.
If we specify execution in this spec too, and we have an "execution plan" then maybe we should also hand over the "execution plan" to an "executor"

**(Query) Processor** - DBMS
Again, a reference to Database systems. The problem with both orchestrator / executor (at least in the previously defined context) is that they both require a definition (plan) of what they have to orchestrate or execute. In the databases theres is a component called the "Query Processor". In  [Encyclopedia of Database Systems](https://link.springer.com/referencework/10.1007/978-0-387-39940-9) it's defined as

> The query processor in a database management system receives as input a query request in the form of SQL text, parses it, generates an execution plan, and completes the processing by executing the plan and returning the results to the client.

**Comment by andimarek** [comment]
I use the term distributed engine for gateway core in Atlassian. 

Because fundamentally it is fully spec conform engine, which does its job by distributing the execution.
